### PR TITLE
Use EMOS coefficients cube

### PR DESCRIPTION
**Pull request by gavinevans**
Addresses #855 

Description
This PR contains changes required to use the coefficients cube added in #850. This also builds on top of #856, which removed redundant functionality that is not required for usage in IMPROVER. The main aim of this PR is to use the coefficients cube within both the estimation of the coefficients and the application of the coefficients. Both of these edits are required in this PR to support the pre-existing EnsembleCalibration plugin and ensemble-calibration CLI, which chain together the estimation of the coefficients and application of the coefficients into one call. However, with a view to real-time running, we would like to separate the estimation and application of the coefficients, so that the coefficients can be estimated ahead of time with the application of the coefficients being a relatively quick process.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 1]
```diff
@@ -372,6 +377,10 @@ def __init__(self, distribution, desired_units,
                     "A default initial guess will be used without "
                     "estimating coefficients from a linear model.")
                 warnings.warn(msg, ImportWarning)
+        else:
+            statsmodels_found = True
+            import statsmodels.api as sm
+            self.sm = sm
         self.statsmodels_found = statsmodels_found
 
     def __str__(self):
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: This needs updating too?

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 2]
```diff
@@ -425,24 +434,50 @@ def create_coefficients_cube(
             raise ValueError(msg)
 
         coefficient_index = iris.coords.DimCoord(
-            np.arange(len(optimised_coeffs)),
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Revert this change?

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 3]
```diff
@@ -425,24 +434,50 @@ def create_coefficients_cube(
...
-            except CoordinateNotFoundError:
-                pass
-        attributes = {"diagnostic_standard_name": current_forecast.name()}
-        for attribute in current_forecast.attributes.keys():
+
+        # Create a forecast_reference_time coordinate.
+        frt_point = cycletime_to_datetime(self.current_cycle)
+
+        try:
+            frt_coord = (
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: What's the benefit of only copying if the coordinate is on the historic cube? Do we want to add these coordinates anyway?

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 4]
```diff
@@ -1018,22 +910,26 @@ def format_calibration_method(calibration_method):
                  "nonhomogeneous gaussian regression"]):
             if (format_calibration_method(self.distribution) in
                     ["gaussian", "truncated gaussian"]):
+                current_cycle = datetime_to_cycletime(
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Should you pass in the current cycle? If not it should be taken from the "forecast_reference_time" coordinate rather than "time"?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 5]
```diff
@@ -587,23 +620,22 @@ def test_coefficient_values_for_gaussian_distribution(self):
         data = [4.55819380e-06, -8.02401974e-09,
                 1.66667055e+00, 1.00000011e+00]
 
-        current_forecast = self.current_temperature_forecast_cube
-
-        historic_forecasts = self.historic_temperature_forecast_cube
+        historic_forecast = self.historic_temperature_forecast_cube
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: Can't you just pass in self.historic_temperature_forecast_cube and self.temperature_truth_cube in rather than doing this?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_ApplyCoefficientsForEnsembleCalibration.py [hunk 6]
```diff
@@ -634,87 +281,16 @@ def test_calibrated_variance_realizations(self):
...
+        _, forecast_variance = plugin._apply_params(
+            predictor_cube, variance_cube)
         self.assertArrayAlmostEqual(forecast_variance.data, data,
                                     decimal=4)
 
-    @ManageWarnings(
-        ignored_messages=["Collapsing a non-contiguous coordinate.",
-                          "invalid escape sequence"],
-        warning_types=[UserWarning, DeprecationWarning])
-    def test_coefficients_realizations(self):
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_ApplyCoefficientsForEnsembleCalibration.py]: Why has this test been removed?

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 7]
```diff
@@ -425,24 +434,50 @@ def create_coefficients_cube(
...
+                historic_forecast.coord("forecast_reference_time").copy(
+                    datetime_to_iris_time(frt_point, time_units="seconds")))
+        except CoordinateNotFoundError:
+            pass
+        else:
+            aux_coords_and_dims.append((frt_coord, None))
+
+        # Create a forecast_period and a time coordinate.
+        try:
+            fp_point, = (
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]:
```suggestion
            copy_of_forecast_period_coord = historic_forecast.coord("forecast_period").copy().convert_units("seconds")
            fp_point, = (np.unique(copy_of_forecast_period_coord.points)))
```

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 8]
```diff
@@ -425,24 +434,50 @@ def create_coefficients_cube(
...
+                np.unique(historic_forecast.coord("forecast_period").points))
+        except CoordinateNotFoundError:
+            pass
+        else:
+            fp_coord = (
+                historic_forecast.coord("forecast_period").copy(fp_point))
+            aux_coords_and_dims.append((fp_coord, None))
+            frt_point = cycletime_to_datetime(self.current_cycle)
+            time_point = (
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Move into if statement and check units of time coord so can put in matching time_units into datetime_to_iris_time

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 3 above)
Comment (fionaRust) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Ignore this comment for now.

File: lib/improver/utilities/temporal.py [hunk 9]
```diff
@@ -71,6 +71,23 @@ def cycletime_to_datetime(cycletime, cycletime_format="%Y%m%dT%H%MZ"):
     return datetime.strptime(cycletime, cycletime_format)
 
 
+def datetime_to_cycletime(adatetime, cycletime_format="%Y%m%dT%H%MZ"):
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/utilities/temporal.py]: I've updated the docstring for this function to help clarify that it is the reverse of the function above.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py [hunk 10]
```diff
@@ -279,9 +279,13 @@ def _create_historic_forecasts(cube, number_of_days=5):
         num=number_of_days, endpoint=True)
     for index in time_range:
         temp_cube = cube.copy()
-        temp_cube.coord("forecast_reference_time").points = (
-            temp_cube.coord("forecast_reference_time").points - index)
-        temp_cube.coord("time").points = temp_cube.coord("time").points - index
+        for coord_name in ["forecast_reference_time", "time"]:
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: After discussion, I've added a TODO statement here, given that the input cube could have units in "seconds" or "hours" depending upon whether the cube used in the unit tests have been generated using the `set_up_variable_cube` functionality or the functionality in the ensemble calibration helper functions. Once the ensemble calibration code has been edited to use set_up_variable_cube more fully, then the _create_historic_forecasts function can be upgraded to work in hours exclusively.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 11]
```diff
@@ -336,6 +337,10 @@ def __init__(self, distribution, desired_units,
             distribution (str):
                 Name of distribution. Assume that the current forecast can be
                 represented using this distribution.
+            current_cycle (str):
+                The current cycle in YYYYMMDDTHHMMZ format e.g. 20171122T0100Z.
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: I've updated this docstring.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 1 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: I've updated this `__str__` to be a `__repr__` and added unit tests. I've also added `__repr__` methods to the other classes in ensemble_calibration.py and added associated unit tests.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 2 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Reverted.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 4 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: You're correct that this should be the `forecast_reference_time` coordinate. I could pass in the `current_cycle` to the ensemble_calibration class, however, in some ways this is duplicating information that is already available.  What do you think?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 5 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: Yes, I've updated this here and for the rest of the unit tests within this file, where possible, so as to avoid an intermediate variable.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 7 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: After discussion, this section has been re-worked something like this.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 8 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Yes, the code specifically relevant to the `time` coordinate has been moved this the if statement.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_ApplyCoefficientsForEnsembleCalibration.py (hunk 6 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_ApplyCoefficientsForEnsembleCalibration.py]: I've removed `test_coefficients` and `test_coefficients_realizations` as I am no longer outputting the coefficients from the `ApplyCoefficientsFromEnsembleCalibration`, as these are now output from `EstimateCoefficientsForEnsembleCalibration`.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 12]
```diff
@@ -431,18 +455,50 @@ def create_coefficients_cube(
...
+                fp_point, = np.unique(copy_of_fp_coord.points)
+                time_point = (
+                    frt_point + datetime.timedelta(seconds=float(fp_point)))
+                time_point = datetime_to_iris_time(
+                    time_point,
+                    time_units=str(historic_forecast.coord("time").units))
+                time_coord = historic_forecast.coord("time").copy(time_point)
+                aux_coords_and_dims.append((time_coord, None))
+
+        attributes = {"diagnostic_standard_name": historic_forecast.name()}
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Have you got a test that checks for this attribute?

Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Yes, I think as a result of doing assertEqual on the result and expected cubes.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 4 above)
Comment (fionaRust) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Yes that's fine, I don't think we need to pass it in.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 13]
```diff
@@ -154,24 +165,68 @@ def test_statsmodels_realizations(self, warning_list=None):
...
+               "desired_units: Kelvin; "
+               "predictor_of_mean_flag: realizations; "
+               "minimiser: <class 'improver.ensemble_calibration."
+               "ensemble_calibration."
+               "ContinuousRankedProbabilityScoreMinimisers'>; "
+               "coeff_names: ['gamma', 'delta', 'alpha', 'beta']>")
+        self.assertEqual(result, msg)
+
+
 class Test_create_coefficients_cube(IrisTest):
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: So we need a test to check what happens if historic forecasts forecast period coord does not all have the same value? Or is this covered elsewhere?

**Comment by fionaRust** [comment]
Suggested change in process method of GeneratePercentilesFromProbabilities line 554:
```
        if not isinstance(forecast_probabilities, iris.cube.Cube):
            forecast_probabilities = forecast_probabilities.concatenate_cubes()
```
All unit tests and CLI tests pass with this change. It stops the var names being lost unnecessarily.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 14]
```diff
@@ -431,18 +455,50 @@ def create_coefficients_cube(
...
+            pass
+        else:
+            aux_coords_and_dims.append((fp_coord, None))
+            if historic_forecast.coords("time"):
+                frt_point = cycletime_to_datetime(self.current_cycle)
+                # Ensure that the fp_point is determined with units of seconds.
+                copy_of_fp_coord = (
+                    historic_forecast.coord("forecast_period").copy())
+                copy_of_fp_coord.convert_units("seconds")
+                fp_point, = np.unique(copy_of_fp_coord.points)
```
Comment (bayliffe) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Didn't we just do lot's of this above? Could we do this once rather than twice with a little restructuring? Or is the difference that we no longer want to change to seconds above but we do here?

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 15]
```diff
@@ -431,18 +455,56 @@ def create_coefficients_cube(
...
+                    datetime_to_iris_time(frt_point, time_units="seconds")))
+        except CoordinateNotFoundError:
+            pass
+        else:
+            aux_coords_and_dims.append((frt_coord, None))
+
+        # Create forecast period and time coordinates.
+        try:
+            fp_point = (
+                np.unique(historic_forecast.coord("forecast_period").points))
```
Comment (bayliffe) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: The previous comments here suggested this needed to be converted to seconds. Why have you chosen not to do this?

File: lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py [hunk 16]
```diff
@@ -551,10 +551,8 @@ def process(self, forecast_probabilities, no_of_percentiles=None,
                 "GeneratePercentilesFromProbabilities")
 
         threshold_coord = find_threshold_coordinate(forecast_probabilities)
-        forecast_probabilities = concatenate_cubes(
-            forecast_probabilities,
-            coords_to_slice_over=threshold_coord,
-            coordinates_for_association=[])
+        if not isinstance(forecast_probabilities, iris.cube.Cube):
+            forecast_probabilities = forecast_probabilities.concatenate_cubes()
```
Comment (bayliffe) [inline lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py]: Is this to address a current use case or is it just a hedge against a cubelist making it here somehow?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 17]
```diff
@@ -376,6 +376,8 @@ def test_alternative_calibration_name(self):
             self.historic_temperature_forecast_cube,
             self.temperature_truth_cube)
         self.assertIsInstance(result, tuple)
+        self.assertIsInstance(result[0], iris.cube.Cube)
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: The doc-string for the test refers to a cubelist, where as you get cubes back (they just happen to be in a tuple right?).

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 13 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: I've added a unit test to cover this.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 18]
```diff
@@ -553,27 +643,22 @@ def setUp(self):
...
         result = plugin.estimate_coefficients_for_ngr(
-            current_forecast, historic_forecasts, truth)
-        self.assertIsInstance(result, dict)
-        self.assertIsInstance(plugin.coeff_names, list)
-        for key in result.keys():
-            self.assertEqual(len(result[key]), len(plugin.coeff_names))
+            self.historic_temperature_forecast_cube,
+            self.temperature_truth_cube)
+        self.assertIsInstance(result, iris.cube.Cube)
+        self.assertArrayAlmostEqual(len(result.data), len(self.coeff_names))
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: These tests don't match what is described in the doc string.

File: lib/improver/tests/utilities/test_temporal.py [hunk 19]
```diff
@@ -77,6 +78,27 @@ def test_define_cycletime_format(self):
...
+        self.assertIsInstance(result, str)
+        self.assertEqual(result, cycletime)
+
+    def test_define_cycletime_format(self):
+        """Test when a cycletime is defined."""
+        dt = datetime.datetime(2017, 11, 22, 1, 0)
+        cycletime = "201711220100"
+        result = datetime_to_cycletime(dt, cycletime_format="%Y%m%d%H%M")
+        self.assertEqual(result, cycletime)
+
```
Comment (bayliffe) [inline lib/improver/tests/utilities/test_temporal.py]: Could you stick in a test demonstrating that it handles seconds (e.g. by ignoring them unless a custom format is included that uses them).

File: lib/improver/utilities/temporal.py [hunk 20]
```diff
@@ -52,16 +52,17 @@
...
      Args:
          cycletime (string):
              A cycletime that can be converted into a datetime using the
              cycletime_format supplied.
 
      Keyword Args:
-         cycletime_format (string):
-             String containg the appropriate directives to indicate how
+         cycletime_format (str):
+             String containing the appropriate directives to indicate how
```
Comment (bayliffe) [inline lib/improver/utilities/temporal.py]: This description is incorrect to my eye. This is specifying the format of the string containing the cycletime that we wish to interpret rather than having anything to do with how we wish it to be displayed.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 14 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Yes, you're right that there is a bit of duplication here but it is for the purpose of being able to ensure that the forecast period coordinate is in seconds here, so that the construction of the `timedelta` object will work as intended, as this is expecting seconds as input: `datetime.timedelta(seconds=float(fp_point))`.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 15 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: This is just for the purpose of supporting inputs with different units for the forecast_period. With the current structuring of the code, the conversion of forecast_period to seconds only occurs where it is required i.e. to support the construction of the `timedelta` object.

File: lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py (hunk 16 above)
Comment (gavinevans) [reply lib/improver/ensemble_copula_coupling/ensemble_copula_coupling.py]: The reason for this particular change is that the IMPROVER concatenate_cubes functionality also strips var names from cubes, as this previously caused some problems in that it was my previous observation that var_names were added upon saving, but were not present prior to saving and hence, to lead to some mismatches in coordinates depending upon they had just been loaded from a file or had been constructed in memory. In this case, stripping the var_names was causing a problem in ApplyCoefficientsFromEnsembleCalibration, as coordinate that should be the same were differing only in their var_names. However as for the bigger picture question, we quite probably do not actually need to support cubelists in this plugin now, so we could do a little bit of a clean-up in a follow-on ticket.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 21]
```diff
@@ -358,10 +377,7 @@ def __init__(self, distribution, desired_units,
...
         import imp
         try:
-            statsmodels_found = imp.find_module('statsmodels')
-            statsmodels_found = True
-            import statsmodels.api as sm
-            self.sm = sm
+            imp.find_module('statsmodels')
         except ImportError:
             statsmodels_found = False
             if predictor_of_mean_flag.lower() in ["realizations"]:
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: I've updated this everywhere that it was used in ensemble_calibration.py.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py (hunk 17 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: I've amended what is returned for this unit test.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 18 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: I've updated the docstrings for these tests.

File: lib/improver/tests/utilities/test_temporal.py (hunk 19 above)
Comment (gavinevans) [reply lib/improver/tests/utilities/test_temporal.py]: I've added a unit test to demonstrate the handling of seconds.

File: lib/improver/utilities/temporal.py (hunk 20 above)
Comment (gavinevans) [reply lib/improver/utilities/temporal.py]: I've updated this description.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 22]
```diff
@@ -272,31 +272,31 @@ def truncated_normal_crps_minimiser(
...
                 Data to be used as the predictor,
                 either the ensemble mean or the ensemble realizations.
-            truth (Numpy array):
+            truth (np.ndarray):
                 Data to be used as truth.
-            forecast_var (Numpy array):
+            forecast_var (np.ndarray):
                 Ensemble variance data.
-            sqrt_pi (Numpy array):
+            sqrt_pi (np.ndarra):
```
Comment (bayliffe) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: missing a "y"

Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Added the missing letter.
